lib/publiccloud/ec2.pm

```diff
@@ -0,0 +1,106 @@
+# SUSE's openQA tests
+#
+# Copyright © 2018 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Helper class for amazon ec2
+#
+# Maintainer: Clemens Famulla-Conrad <cfamullaconrad@suse.de>
+
+package publiccloud::ec2;
+use Mojo::Base 'publiccloud::provider';
+use testapi;
+
+has ssh_key      => undef;
+has ssh_key_file => undef;
+
+sub init {
+    my ($self, %params) = @_;
+
+    assert_script_run("export AWS_ACCESS_KEY_ID=" . $self->key_id);
+    assert_script_run("export AWS_SECRET_ACCESS_KEY=" . $self->key_secret);
+    assert_script_run('export AWS_DEFAULT_REGION="' . $self->region . '"');
+}
+
+sub find_img {
+    my ($self, $name) = @_;
+
+    $name = $self->prefix . '-' . $name;
+
+    my $out = script_output("aws ec2 describe-images  --filters 'Name=name,Values=$name'");
+    if ($out =~ /"ImageId":\s+"([^"]+)"/) {
+        return $1;
+    }
+    return;
+}
+
+sub create_ssh_key {
+    my ($self, $prefix, $out_file) = @_;
+
+    return $self->ssh_key if ($self->ssh_key);
+
+    for my $i (0 .. 9) {
+        my $key_name = $prefix . "_" . $i;
+        my $cmd      = "aws ec2 create-key-pair --key-name '" . $key_name
+          . "' --query 'KeyMaterial' --output text > " . $out_file;
+        my $ret = script_run($cmd);
+        if (defined($ret) && $ret == 0) {
+            $self->ssh_key($key_name);
+            $self->ssh_key_file($out_file);
+            return $key_name;
+        }
+    }
+    return;
+}
+
+sub delete_ssh_key {
+    my $self = shift;
+    my $name = shift || $self->ssh_key;
+
+    return unless $name;
+
+    assert_script_run("aws ec2 delete-key-pair --key-name " . $name);
+    $self->ssh_key(undef);
+}
+
+sub upload_img {
+    my ($self, $file) = @_;
+
+    die("Create key-pair failed") unless ($self->create_ssh_key($self->prefix . time, 'QA_SSH_KEY.pem'));
+
+    my ($img_name) = $file =~ /([^\/]+)$/;
+
+    assert_script_run("ec2uploadimg --access-id '"
+          . $self->key_id
+          . "' -s '"
+          . $self->key_secret . "' "
+          . "--backing-store ssd "
+          . "--grub2 "
+          . "--machine 'x86_64' "
+          . "-n '" . $self->prefix . '-' . $img_name . "' "
+          . (($img_name =~ /hvm/i) ? "--virt-type hvm --sriov-support " : "--virt-type para ")
+          . "--verbose "
+          . "--regions '" . $self->region . "' "
+          . "--ssh-key-pair '" . $self->ssh_key . "' "
+          . "--private-key-file " . $self->ssh_key_file . " "
+          . "-d 'OpenQA tests' "
+          . "'$file'",
+        timeout => 60 * 60
```

Comment:
I vote for defining parent class parameter `timeout` and give value in each children

Reply:
I'm not sure. The timeout depends on the provider and how images gets uploaded. I do also not see to define a generic timeout, when doing so, then it should be a specific upload_image_timeout.
What I see here is, give the timeout as parameter to the upload image function and if not given the default from the specific implementation.

Reply:
you misunderstood my idea, but let's skip it for now